autoprecompiles/src/execution/candidates.rs

```diff
@@ -0,0 +1,567 @@
+use std::cmp::Ordering;
+use std::sync::Arc;
+
+use itertools::Itertools;
+
+use crate::execution::{ExecutionState, OptimisticConstraintEvaluator, OptimisticConstraints};
+
+pub struct ApcCandidates<E: ExecutionState, A, S> {
+    candidates: Vec<ApcCandidate<E, A, S>>,
+}
+
+impl<E: ExecutionState, A, S> Default for ApcCandidates<E, A, S> {
+    fn default() -> Self {
+        Self {
+            candidates: Default::default(),
+        }
+    }
+}
+
+impl<E: ExecutionState, A: Apc, S: Snapshot> ApcCandidates<E, A, S> {
+    /// Given the current state of execution, retain the candidates whose constraints are still
+    /// verified
+    pub fn check_conditions(
+        &mut self,
+        state: &E,
+        snapshot_callback: impl Fn() -> S,
+    ) -> Vec<ApcCandidate<E, A, S>> {
+        let mut snapshot = None;
+
+        // Filter out failing candidates and upgrade the ones that are done
+        self.candidates
+            .retain_mut(|candidate| match &mut candidate.status {
+                // Check the conditions for unconfirmed candidates
+                CandidateStatus::InProgress(optimistic_constraint_evaluator) => {
+                    if optimistic_constraint_evaluator.try_next(state).is_err() {
+                        return false;
+                    }
+                    match candidate.remaining_instr_count == 0 {
```

Comment:
when exactly is `check_conditions` is called?
if its called _after_ each instruction is executed, then maybe we should decrement and then check instead?
if its called _before_ each instruction, then the pending APCs at the last instructions won't have a chance to be `Done`?